### PR TITLE
Fix train.py for MSE sorting

### PR DESCRIPTION
**Pull request by seldauyanik-maxim**
Performance metrics history dictionary (perf_scores_history) was being sorted wrt MSE value in reverse order and the largest one was being selected as best. However, for the MSE case (for regression problems) lowest MSE should be selected.

Top1 score is replaced with 1-MSE for regression problems for correctly selecting and saving the proper model.

### CODE REVIEW

**Comment by alicangok** [comment]
Would continuing to log the regular MSE, but only changing the two associated sorting lines, by setting "reverse=False" also work? Logging "1-MSE" seemed counter-intuitive to me.

**Comment by seldauyanik-maxim** [comment]
> Would continuing to log the regular MSE, but only changing the two associated sorting lines, by setting "reverse=False" also work? Logging "1-MSE" seemed counter-intuitive to me.

Indeed seems more counter-intuitive however, when sorting not only MSE is used but coupled with other keys that seems to need to have higher values for a better case.

**Comment by alicangok** [comment]
> > Would continuing to log the regular MSE, but only changing the two associated sorting lines, by setting "reverse=False" also work? Logging "1-MSE" seemed counter-intuitive to me.
> 
> Indeed seems more counter-intuitive however, when sorting not only MSE is used but coupled with other keys that seems to need to have higher values for a better case.

Ah, you're right. Perhaps we could do multi-pass sorts like below, although that's very elegant.

```
            # Sort by MSE as main sort key, then sort by epoch
            perf_scores_history.sort(key=operator.attrgetter('epoch'), reverse=True)
            perf_scores_history.sort(key=operator.attrgetter('top1'), reverse=False)
            if args.sparsity_perf:
                # Sort by sparsity as main sort key
                perf_scores_history.sort(key=operator.attrgetter('params_nnz_cnt'), reverse=True)
```

**Comment by seldauyanik-maxim** [comment]
> > > Would continuing to log the regular MSE, but only changing the two associated sorting lines, by setting "reverse=False" also work? Logging "1-MSE" seemed counter-intuitive to me.
> > 
> > 
> > Indeed seems more counter-intuitive however, when sorting not only MSE is used but coupled with other keys that seems to need to have higher values for a better case.
> 
> Ah, you're right. Perhaps we could do multi-pass sorts like below, although that's not very elegant.
> 
> ```
>             # Sort by MSE as main sort key, then sort by epoch
>             perf_scores_history.sort(key=operator.attrgetter('epoch'), reverse=True)
>             perf_scores_history.sort(key=operator.attrgetter('top1'), reverse=False)
>             if args.sparsity_perf:
>                 # Sort by sparsity as main sort key
>                 perf_scores_history.sort(key=operator.attrgetter('params_nnz_cnt'), reverse=True)
> ```

That also makes sense. Btw top1 is used for accuracy like performance metrics throughout train.py and we also have an objective loss. For the regression case, they are reported as equal. I guess, for some similarity 1-MSE was being in use for some logging purposes.
I agree with both versions. Lets get some more votes on this and I can implement accordingly.

**Comment by oguzhanbsolak** [comment]
[2022.11.14-191711.log](https://github.com/MaximIntegratedAI/ai8x-training/files/10010367/2022.11.14-191711.log)
I tried this fix with the faceID model, and it seems working as intended. You may find the log file attached. However, my vote is for the multipass solution as it's less confusing for the users.

File: train.py [hunk 1]
```diff
@@ -1311,16 +1311,16 @@ def update_training_scores_history(perf_scores_history, model, top1, top5, mAP,
 
             # Keep perf_scores_history sorted from best to worst
             if not args.sparsity_perf:
-                # Sort by mse as main sort key, then sort by epoch
+                # Sort by mse (1 - mse) as main sort key, then sort by epoch
```
Comment (ermanok) [inline train.py:1314]: It seems, the code assumes to MSE be in range [0, 1] but it might not be. -MSE might be a better choice.

**Review by ermanok — COMMENTED** [review]
-MSE might be better to use if MSE is not guaranteed to be in [0, 1]

**Comment by seldauyanik-maxim** [comment]
Dear all,
All above and the latest comments are addressed in the last commit
* top1 score is assigned to MSE directly (instead of -MSE or 1-MSE)
* sequential sorting implemented in perf scores history
* 1-MSE parts totally removed

**Review by ermanok — APPROVED** [review]
Looks good.

**Pull request merged**